### ISSUE TITLE
Unable to use broadcast - getting 503 response

### ISSUE THREAD

**Issue by jNullj** [ampache/ampache#1769]
# Environment

 * Ampache version: 3.8.8 (and just now tested with 3.8.9 and its the same)
 * Web server + version: apache 2.4.34
 * Server operating system: archlinux 64bit
 * Client operating system: archlinux 64bit , windows 10 64bit
 * using ssl and ssl authentication
 * libevent 2.1.8

 * Client type:
   - Web
   - Ampache

# Settings

in /config/ampache.cfg.php:
``` websocket_address = "wss://sub.domain.com/ws" ```

in apache vhost file:
```
ServerName sub.domain.com
...
ProxyPass /ws ws://sub.domain.com:8101/ retry=0
ProxyPassReverse /ws ws://sub.domain.com:8101/ retry=0
ProxyRequests off
ProxyTimeout 15
```

# Description

Tried using broadcast, but it seems like im getting 503 response on every time, playing with the configuration i was able to get 404 once, no idea what i did.
Doesnt work localy on the server as well.
while using ws im getting connection refused (might have to do with port 80 blocked in the firewall) so i tried moving to wss.

Client's request header:
```
GET wss://sub.domain.com/ws/broadcast HTTP/1.1
Host: sub.domain.com
Connection: Upgrade
Pragma: no-cache
Cache-Control: no-cache
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36
Upgrade: websocket
Origin: https://sub.domain.com
Sec-WebSocket-Version: 13
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US;q=0.8,en;q=0.7
Cookie: cookie_disclaimer=removed; jp_volume=removed; ampache=removed; ampache_user=removed; ampache_lang=en_US
Sec-WebSocket-Key: removed
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
```

Client's response:
```
HTTP/1.1 503 Service Unavailable
Date: Tue, 11 Sep 2018 08:22:58 GMT
Server: Apache/2.4.34 (Unix) OpenSSL/1.1.0i
Vary: accept-language,accept-charset
Accept-Ranges: bytes
Connection: close
Content-Type: text/html; charset=utf-8
Content-Language: en
```

websocket_run crush on the attempt with:
```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Broadcast_Server::notifyNbListeners() must be an instance of Broadcast, null given, called in /usr/share/webapps/ampache/lib/class/broadcast_server.class.php on line 299 and defined in /usr/share/webapps/ampache/lib/class/broadcast_server.class.php:353
Stack trace:
#0 /usr/share/webapps/ampache/lib/class/broadcast_server.class.php(299): Broadcast_Server->notifyNbListeners(NULL)
#1 /usr/share/webapps/ampache/lib/class/broadcast_server.class.php(105): Broadcast_Server->registerListener(Object(Ratchet\WebSocket\Version\RFC6455\Connection), '19')
#2 /usr/share/webapps/ampache/lib/vendor/cboden/ratchet/src/Ratchet/WebSocket/Version/RFC6455.php(204): Broadcast_Server->onMessage(Object(Ratchet\WebSocket\Version\RFC6455\Connection), 'REGISTER_LISTEN...')
#3 /usr/share/webapps/ampache/lib/vendor/cboden/ratchet/src/Ratchet/WebSocket/Version/RFC6455.php(208): Ratchet\WebSocket\Version\RFC6455->onMessage(Object(Ratchet\WebSocket\Version\RFC6455\Connection), '\x81\x95m\x93\x96\xE1? in /usr/share/webapps/ampache/lib/class/broadcast_server.class.php on line 353
```

before the crush i get 101 response on client's side~
```
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: removed
X-Powered-By: Ratchet/0.3.6
```

# Logs

apache log's tail:
```
[Tue Sep 11 11:54:45.460349 2018] [proxy:error] [pid 24353:tid 140164110219008] (111)Connection refused: AH00957: WS: attempt to connect to 127.0.0.1:8101 (sub.domain.com) failed
[Tue Sep 11 11:54:45.460377 2018] [proxy_wstunnel:error] [pid 24353:tid 140164110219008] [client 127.0.0.1:49112] AH02452: failed to make connection to backend: sub.domain.com
[Tue Sep 11 11:54:48.365525 2018] [proxy:error] [pid 24565:tid 140163808212736] (111)Connection refused: AH00957: WS: attempt to connect to 127.0.0.1:8101 (sub.domain.com) failed
[Tue Sep 11 11:54:48.365557 2018] [proxy_wstunnel:error] [pid 24565:tid 140163808212736] [client 127.0.0.1:49148] AH02452: failed to make connection to backend: sub.domain.com
[Tue Sep 11 11:54:50.407496 2018] [proxy:error] [pid 24352:tid 140164192233216] (111)Connection refused: AH00957: WS: attempt to connect to 127.0.0.1:8101 (sub.domain.com) failed
[Tue Sep 11 11:54:50.407524 2018] [proxy_wstunnel:error] [pid 24352:tid 140164192233216] [client 127.0.0.1:49170] AH02452: failed to make connection to backend: sub.domain.com
[Tue Sep 11 11:54:51.762155 2018] [proxy:error] [pid 24353:tid 140164068255488] (111)Connection refused: AH00957: WS: attempt to connect to 127.0.0.1:8101 (sub.domain.com) failed
[Tue Sep 11 11:54:51.762185 2018] [proxy_wstunnel:error] [pid 24353:tid 140164068255488] [client 127.0.0.1:49184] AH02452: failed to make connection to backend: sub.domain.com
[Tue Sep 11 11:54:53.062874 2018] [proxy:error] [pid 24352:tid 140164209018624] (111)Connection refused: AH00957: WS: attempt to connect to 127.0.0.1:8101 (sub.domain.com) failed
[Tue Sep 11 11:54:53.062913 2018] [proxy_wstunnel:error] [pid 24352:tid 140164209018624] [client 127.0.0.1:49198] AH02452: failed to make connection to backend: sub.domain.com
```

**Comment by wagnered** [comment]
@jNullj 

I'll be honest with ya... I have never tested the broadcast function of Ampache.  It is not working on my system, but seemingly for different reasons.

From the PHP fatal error, It looks like there is no broadcast server created or the broadcast _id is not set correctly.

I'm not sure why your web socket address is set as an internet URL instead of localhost (127.0.0.1).  If the ampache server is on the same system as the apache server, a remote URL shouldn't be necessary. 

I  ran 'php bin/websocket_run.inc -p 8100` ( without the ` > \/dev\/null &`).  I get a series of messages like:

>[1536714136][info]Broadcast 11 has song position to 123.158639.

on the cli.

But the browser console indicates that Firefox can't connect to the socket.  I probably have a rewrite/redirect problem with Apache or Ampache. :frowning: I''m guessing that this is at least part of your problem.

**Comment by jNullj** [comment]
After changing in /config/ampache.cfg.php to:
` websocket_address = "ws://localhost/ws" `

and vhosts to:
```
ServerName sub.domain.com
...
ProxyPass /ws ws://localhost:8101/ retry=0
ProxyPassReverse /ws ws://localhost:8101/ retry=0
ProxyRequests off
ProxyTimeout 15
```
im getting response 403

**Comment by wagnered** [comment]
@jNullj 

> im getting response 403

Indicates forbidden access. Do you get that response on both the linux and windows system?  If from Windows, you will need to use your network address, (mine being 192.168.2.3) instead of localhost.

I have identified a problem with the code and am working on the fix. With the current change, when I try listening to a broadcast from a user login, I see the output of `php bin/websocket_run.inc -v -p 8100` displaying the current song position, but I don't hear the song or see the album image. That means that the websocket is working but the music stream might be redirected correctly.

I'll report back on progress.

**Comment by jNullj** [comment]
After changing all localhost to the domain name, i get a 404 response from ratchet.
debugging a bit i found ratchet respones with 404 after it gets exception (no match) at the routing match function ([here](https://github.com/ratchetphp/Ratchet/blob/master/src/Ratchet/Http/Router.php#L47))
```php
        try {
            $route = $this->_matcher->match($uri->getPath());
        } catch (MethodNotAllowedException $nae) {
            return $this->close($conn, 405, array('Allow' => $nae->getAllowedMethods()));
        } catch (ResourceNotFoundException $nfe) {
            return $this->close($conn, 404); //running into this part of the code
        }
```

after that i tried to `var_dump($_matcher)` for the condition and also `$request->getPath()`.
Both looks like they fit.

Im puzzeled about the reason it thinks that the route doesnt exists.

the problem seems to be related to symfony, cant put my finger on it yet.
I might try testing with newer version of Ratchet/Symfony.

**Comment by wagnered** [comment]
Maybe enabling php debug logging might reveal something.

**Comment by wagnered** [comment]
@jNullj 

Maybe  issue #1140 will help.

**Comment by jNullj** [comment]
nice, got:
```
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: pcfka8VwRGyo6CrF60Y59ooaL/Q=
X-Powered-By: Ratchet/0.3.6
```

But after that the websocket_server crushed with:
```php
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Broadcast_Server::notifyNbListeners() must be an instance of Broadcast, null given, called in /usr/share/webapps/ampache/lib/class/broadcast_server.class.php on line 299 and defined in /usr/share/webapps/ampache/lib/class/broadcast_server.class.php:353
Stack trace:
#0 /usr/share/webapps/ampache/lib/class/broadcast_server.class.php(299): Broadcast_Server->notifyNbListeners(NULL)
#1 /usr/share/webapps/ampache/lib/class/broadcast_server.class.php(105): Broadcast_Server->registerListener(Object(Ratchet\WebSocket\Version\RFC6455\Connection), '19')
#2 /usr/share/webapps/ampache/lib/vendor/cboden/ratchet/src/Ratchet/WebSocket/Version/RFC6455.php(204): Broadcast_Server->onMessage(Object(Ratchet\WebSocket\Version\RFC6455\Connection), 'REGISTER_LISTEN...')
#3 /usr/share/webapps/ampache/lib/vendor/cboden/ratchet/src/Ratchet/WebSocket/Version/RFC6455.php(208): Ratchet\WebSocket\Version\RFC6455->onMessage(Object(Ratchet\WebSocket\Version\RFC6455\Connection), '\x81\x95\xD5\xA6\xFA)\ in /usr/share/webapps/ampache/lib/class/broadcast_server.class.php on line 353
```

**Comment by wagnered** [comment]
@jNullj 

I'm not sure the broadcast ever worked as it should. You might want to try setting up a channel on a playlist until I can get this worked out (hopefully I can :frowning_face: ). It's not behaving like I expected it to  and unravelling the mashup of php and Javascript will take some time.

**Comment by jNullj** [comment]
Solved the issue, it still crushes but only while connecting to an offline broadcast.
I changed http_host and force_ssl and it worked but only after clearing browser cache which confused me at start.
The only things that doesnt work yet is the album's art. but im fine with that for now.

Thanks alot for the help!
hope someone with the same issue might find this usefull in the future.

**Comment by wagnered** [comment]
I haven't been able to  succeed a second time and I'm not sure exactly what you have done do get the broadcast working to the extent you have.

Could you  document the changes you made to get your broadcast working?    I can debug the album art problem.

I am working on several other issues and your help would be much appreciated.

**Comment by jNullj** [comment]
Some info in the hope i can get anyone solve this

a note:
im using Apache and the server is configured to use https only, all http connections are refused.
also my socket is on port 8101

my vhost file:
```
<VirtualHost *:443>
    ServerAdmin removed
    DocumentRoot "removed"
    ServerName mydomain.com
    ErrorLog "removed/mydomain.com.error.log"
    CustomLog "removed/mydomain.com.common.log" common

    RewriteEngine On
    RewriteRule ^/ws/(.*) ws://mydomain.com:8101/$1 [P]
    ProxyPass "/ws/" "ws://mydomain.com:8101/" retry=0
    ProxyPassReverse "/ws/" "ws://mydomain.com:8101/" retry=0
    ProxyRequests off
    ProxyTimeout 15

    SSLEngine on
    removed keys location
    # SSL auth clients - for post network setup
    removed client authentication location
    SSLVerifyClient require
    SSLVerifyDepth 1
</VirtualHost>
```

running the socket server with:
```sh
php websocket_run.inc -p 8101 -v
```

i changed websocket_run.inc content to the following (its part of the file):
```php
$urlinfo = parse_url(AmpConfig::get('websocket_address'));
$host = $urlinfo['host'];
if (empty($host)) {
    $host = "localhost";
}
$host = "mydomain.com";
$Origins = array("mydomain.com", "127.0.0.1", "localhost");

$app = new Ratchet\App($host, $port, '0.0.0.0');
$brserver = new Broadcast_Server();
$brserver->verbose = $verbose;
//$app->route('/ws/broadcast', $brserver);
var_dump($app->route('/broadcast', $brserver, $Origins));
```

in the ampache config file:
```php
http_host = "mydomain.com"
http_port = 443
web_path = ""
local_web_path = "https://mydomain.com/"
session_cookiesecure = 1
broadcast = "true"
websocket_address = "wss://mydomain.com/ws"
force_ssl = "true"
```

**Comment by wagnered** [comment]
@jNullj 

I noticed an apparently problematic setting:

In the ampache.cfg.php file, you have websocket_access set to secure socket(wss) but
in the vhost setting it appears to be set to insecure socket(ws).  Is this correct?

Also, changing to $origins from array(*) in "websocket_run.inc" just narrows the path(s) down to specific url's.

**Comment by wagnered** [comment]
When I run `bin/websocket_run.inc -v -p 8100` I can see that the player gets registered and it displays the elapsed time on the cli.  But I cannot get a listener registered.

I log in as another user and go to "Broadcasts"  then start the broadcast.  The cli displays the message
>Wrong listener session <session id>

Stepping through the logic, I see Ampache requires a stream type session. The interface login id is given instead of the stream id.

I wonder if I'm trying to start listening the wrong way or a setting might be wrong.

**Comment by jNullj** [comment]
im glad you was able to get that working
i use unsecured socket at the proxy becouse ratchet (to my understanding) does not support secured socket, so insted apache's proxy redirects into unsecured socket internaly.
But i still find it odd how thats works, i never checked the communication is encrypted, i might will with wireshark later.
but considering the information that passes throw the socket, its not top priority.

i have a few bugs i might open tickets in the future.

**Comment by wagnered** [comment]
I finally got it working(mostly).  I am getting the music, album art and album/artist info on Firefox for Android.
I also fixed a bug that caused a problem with the URL if `web_path` is set. I'll be pushing the change to the develop branch.

I still don't have it working with "URL rewrite"  (stream_beautiful_url) enabled.  For some reason, Apache's mod_rewrite isn't working correctly.

I added:
```
<Directory "/path/to/play folder">
    AllowOverride All
 </Directory>
```
to the vhost but it didn't  work.
